include/RED4ext/Relocation-inl.hpp

```diff
@@ -27,50 +26,207 @@ uintptr_t RED4ext::UniversalRelocBase::Resolve(uint32_t aHash)
-                   << ".\nProcess will now close.";
+            std::wostringstream stream;
+            stream << L"Failed to find the address for the hash (" << std::dec << aHash << ") provided by the plugin.\n"
+                   << L"This issue is likely caused by the mod using an incorrect or outdated hash.";
 
-            MessageBoxA(nullptr, stream.str().c_str(), "RED4ext.SDK", MB_ICONERROR | MB_OK);
-            TerminateProcess(GetCurrentProcess(), 1);
+            ShowErrorAndTerminateProcess(stream.str(), 0);
         }
 
         return address;
     }
 }
+
+RED4EXT_INLINE HMODULE RED4ext::UniversalRelocBase::GetRED4extModule()
+{
+    constexpr auto moduleName = L"RED4ext.dll";
+
+    auto handle = GetModuleHandleW(moduleName);
+    if (!handle)
+    {
+        auto msg = L"The mod you are using could not locate the necessary module (i.e. RED4ext.dll) in the "
+                   L"loaded modules, which is required by the mod to resolve addresses correctly.\n"
+                   L"This may occur if RED4ext is not properly loaded into the current process.\n"
+                   L"\n"
+                   L"Please ensure that RED4ext is correctly installed.\n"
+                   L"\n"
+                   L"If you are the mod's developer, verify that your mod was loaded by RED4ext. "
+                   L"Alternatively, you may need to provide your own address resolver.";
+
+        ShowErrorAndTerminateProcess(msg, GetLastError());
+    }
+
+    return handle;
+}
+
+RED4EXT_INLINE RED4ext::UniversalRelocBase::ResolveFunc_t RED4ext::UniversalRelocBase::
+    InitializeAddressResolverFunction()
+{
+    constexpr auto procName = "RED4ext_ResolveAddress";
+
+    auto handle = GetRED4extModule();
+
+    auto func = reinterpret_cast<ResolveFunc_t>(GetProcAddress(handle, procName));
+    if (func == nullptr)
+    {
+        auto msg = L"The mod you are using is unable to find the required address resolver function from RED4ext.\n"
+                   L"This may occur if RED4ext is not properly loaded, OR if the mod is incompatible with the current "
+                   L"version of RED4ext.\n"
+                   L"\n"
+                   L"Please ensure that RED4ext is correctly installed AND that both RED4ext and the mod are "
+                   L"up-to-date.";
+
+        ShowErrorAndTerminateProcess(msg, GetLastError());
+    }
+
+    return func;
+}
+
+RED4EXT_INLINE RED4ext::UniversalRelocBase::ResolveFunc_t RED4ext::UniversalRelocBase::GetAddressResolverFunction()
+{
+    static const ResolveFunc_t func = InitializeAddressResolverFunction();
+    return func;
+}
+
+RED4EXT_INLINE HMODULE RED4ext::UniversalRelocBase::GetCurrentModuleHandle()
+{
+    HMODULE result;
+
+    if (!GetModuleHandleExW(GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS | GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT,
+                            reinterpret_cast<LPCWSTR>(UniversalRelocBase::Resolve), &result))
```

Comment:
would consider changing this to std::bit_cast as reinterpret_cast is not really defined for anything else than cast to char* by the standard, this is relying on undefined behavior atm

Reply:
Sam as https://github.com/wopss/RED4ext.SDK/pull/163#discussion_r1886774809, if nothing changed then func pointer to object pointer should be implementation defined.

Reply:
how do i resolve this? 😅 dont see any resolve button 🤔

Reply:
I changed this to do `bit_cast`, thanks again!

Imo it is better to use a standard defined operation than hope the implementation does what you want.